Make `Environment.ensure()` call from thread-safe context

lorban opened the pull request:
`Environment.ensure()` is explicitly called from `WebAppClassLoaderTest`'s `@BeforeEach` via `TestableWebAppClassLoaderContext.initEnvironment()`, i.e.: when surefire executes tests in parallel, the non-thread-safe `Environment.ensure()` method gets executed by multiple threads in parallel.

Removing `TestableWebAppClassLoaderContext.initEnvironment()` and moving its code to a `TestableWebAppClassLoaderContext` static block makes sure that `Environment.ensure()` is only going to be called only once with the same result.

Fixes #12853

lorban committed:
#12853 make Environment.ensure() call from thread-safe context

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented:
There's 2 ways I would approach this.

1. Flag/Annotate the tests in here that manipulate the Environment as junit isolated so that they do not conflict with each other.  (see past fixes in this regard from @olamy )

2. Embrace the chaos, and just make the Environment name equivalent to the junit TestInfo, which is unique per test case. (easy to setup, not sure how easy it is to troubleshoot)

joakime commented:
Note: this is essentially a duplicate of #12851

janbartel commented:
Can we please make `Environment` threadsafe?

lorban committed:
#12853 make Environment thread-safe

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@janbartel done

janbartel commented on jetty-core/jetty-ee/src/test/java/org/eclipse/jetty/ee/TestableWebAppClassLoaderContext.java:
```diff
@@ -25,19 +25,19 @@
-    private List<Resource> extraClasspath;
-
-    public static void initEnvironment()
+    static
```
I'm not as keen on this change. If this is now static then an individual test case can't use this class differently.

lorban replied:
So far, this class is only used by one test class which always calls `initEnvironment()` in a thread-unsafe way. IMHO we should postpone that static block refactoring to the day we start using `TestableWebAppClassLoaderContext` differently.

janbartel commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java:
```diff
@@ -17,7 +17,7 @@
 import java.util.Collections;
 import java.util.Map;
-import java.util.TreeMap;
+import java.util.concurrent.ConcurrentSkipListMap;
```
Now this is a change I like! :)

lorban replied:
I kinda agree that `Environment`'s static API should have a thread-safe implementation, but the API itself only displays non-thread-safe behaviors.

We better be careful when using the `set(Environment)` call.

sbordet replied:
@lorban I think `Environment.set()` should be removed.

It's easily replaced by the constructor of `Named` putting itself in the `Named.__environments` map.

Also, the `Named` constructor (and likely the whole class) should be private, as we don't want to create `Named` instances that are not in the `Named.__environments` map.
It would be enough to have `Environment.ensure(name, ClassLoader)`.

janbartel reviewed (changes requested):
Maybe reconsider the necessity to make the assignment of the variables in the `TestableWebAppClassLoader` static - this may have unintended consequences for other tests?  But +1 for the thread-safety of `Environment`.

lorban committed:
#12853 remove Environment.put() static from API, add Environment.ensure(name, ClassLoader) instead and make Environment.Named package-protected

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@sbordet I replaced `Environment.put()` with `Environment.ensure(name, ClassLoader)` but I'm not sure if that's any better as when `ensure(name, ClassLoader)` does not compute (i.e.: there already is an environment with that name) we should probably compare the classloaders... which kinda is a can of worms.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/NamedEnvironment.java:
```diff
@@ -0,0 +1,54 @@
+import org.eclipse.jetty.util.Attributes;
+import org.eclipse.jetty.util.TypeUtil;
+
+class NamedEnvironment extends Attributes.Mapped implements Environment, Dumpable
```
I don't really like creating a new top level class for such a small trivial thing.  What was wrong with Environment.Named as a package or private inner class?

lorban replied:
The class was implicitly public with no way to alter its visibility as it was enclosed by an interface.

Maybe `Environment` should be a class?

gregw replied:
hmmm or just moved NamedEnvironment to an internal sub package

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java:
```diff
@@ -32,22 +28,22 @@ public interface Environment extends Attributes
     }
 
-    static Environment set(Environment environment)
+    static Environment ensure(String name, ClassLoader classLoader)
```
This should be a `Supplier<ClassLoader>` argument, as ensure semantics should only construct a classloader if the environment doesn't already exist.

lorban replied:
If `EnvironmentBuilder.build()` is supposed to throw when an environment with such name already exists, there is no need to use a supplier for the classloader.

gregw replied:
one or the other

gregw commented on jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/EnvironmentBuilder.java:
```diff
@@ -61,6 +60,6 @@ public void addClassPath(String... classPaths)
     public Environment build() throws Exception
     {
-        return new Environment.Named(_name, new URLClassLoader(_classpath.toArray(new URL[0]), EnvironmentBuilder.class.getClassLoader()));
+        return Environment.ensure(_name, new URLClassLoader(_classpath.toArray(new URL[0]), EnvironmentBuilder.class.getClassLoader()));
```
What if the name already exists? This will just discard the classloader 
We need at least an error

lorban replied:
`ensure` does not allow that. This doesn't look like it's the right contract for the job here.

Let me think of something else...

gregw reviewed (changes requested):
Sorry but if we are going to "clean this up", it needs a bit more work.

lorban committed:
#12853 rework Environment static API + add javadoc + adapt tests to make sure they can be run in parallel

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw I've reworked the `Environment` static API, this time adding javadoc to make sure the contract is clear. It now throws exceptions when the expectations aren't met, so I modified the tests accordingly to make them parallelizable according to this new contract.

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java:
```diff
@@ -30,24 +26,58 @@ public interface Environment extends Attributes
+     * @return the environment
+     * @throws IllegalStateException if an environment with the given name but a non-default classloader already exists
+     */
+    static Environment ensure(String name) throws IllegalStateException
```
I find `ensure()` confusing.

Environments are statics, and are created exactly at the right point by the right classes (the various eeN `ContextHandler`s), and when they do, it's a "create" operation because nobody could have created them.

I think we don't want other code to create environments, other than those eeN `ContextHandler`s.

If that's true, then all the rest of the code should use `get()`, because that code should not create environments as an unexpected side effect, just expect them to exist and use them.

Tests can easily call `create()` and then `get()`.

Am I missing something?

sbordet replied:
@gregw we should discuss how embedded code will be able to deploy different eeN contexts, because I feel that being able to do that will solve the problem with the `Environment` API.

lorban replied:
We need a test that deploys both a ee9 servlet _and_ a ee10 servlet on the same embedded server, as the current modifications break that use-case since the `ensure()` call in `ServletContextHandler`  would detect that the environment is present but with a non-default classloader and throw.

This should probably be part of the documentation too as this isn't a trivial setup.

gregw replied:
i think this PR has got way out of scope for its original intent.
How about we just roll back to making the map concurrent then:
 + perhaps rename `set` to `add` or `create`?
 + make `set` (or whatever it is called) throw an IAE if the named environment already exists.

lorban committed:
revert to previous contract for ensure()

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanups

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Improved ensure

gregw commented:
@lorban I've pushed an update to `Environment.ensure` along the lines discussed with @sbordet this morning.  I've also added an `EnvironmentTest`

gregw committed:
Fixed copyright

sbordet commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/component/EnvironmentTest.java:
```diff
@@ -0,0 +1,66 @@
+        ClassLoader loaderAX = new URLClassLoader(new URL[0]);
+        assertThrows(IllegalStateException.class, () -> Environment.create("envA", loaderAX));
+
+        System.err.println(Environment.getAll());
```
Leftover?

sbordet reviewed (approved):
LGTM besides a leftover System.err.

gregw committed:
update from review

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java:
```diff
@@ -13,41 +13,68 @@
     {
-        return Named.__environments.computeIfAbsent(name, Named::new);
+        Environment environment = NamedEnvironment.ENVIRONMENTS.computeIfAbsent(name, n -> new NamedEnvironment(n, classLoader));
+        if (environment.getClassLoader() != classLoader)
```
So this check means that `EnvironmentBuilder` and `XmlConfiguration` must not configure `ee*` environments otherwise the `ensure()` call in the multiple `ServletContextHandler`s would make those classes fail to load because of this check.

Did I get this right?

gregw replied:
Not at all.   The EB and XC can configure ee* environments, so long as they do so in a way that classes that call ensure are loaded from the classloaders in the created environments

lorban replied:
ok, thanks for the explanation.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java:
```diff
@@ -13,41 +13,68 @@
+     *
+     * @param name the environment name
+     * @return the environment
+     * @throws IllegalStateException if an environment with the given name but a non-default classloader already exists
```
There is no concept of default classloader anymore.

joakime replied:
The current javadoc for Environment.getClassLoader makes it seem that there is still a default classloader.  or at least a fallback classloader.

gregw replied:
updated javadoc

gregw committed:
updates from review

joakime commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletContextHandler.java:
```diff
@@ -133,7 +133,7 @@
 {
     private static final Logger LOG = LoggerFactory.getLogger(ServletContextHandler.class);
-    public static final Environment ENVIRONMENT = Environment.ensure("ee11");
+    public static final Environment ENVIRONMENT = Environment.ensure("ee11", ServletContextHandler.class.getClassLoader());
```
How is this possible?

The Environment is the classloader for the ServletContextHandler.
So if we reach this point, then ...

1. The Environment for "ee11" exists already.
2. The Environment classloader for "ee11" exists already too.

Right?

So why reference the ServletContextHandler.class.getClassLoader at all?

gregw replied:
Exactly! Either:
 a) The environment already exists (probably created by XML), in which case ensure checks that the classloader is the same one; OR
 b) The environment does not exist (embedded usage), in which case the environment is created with the loader that loaded the environment specific class.

sbordet replied:
Saves creating the environments explicitly.

Consider:

```java
ClassLoader serverCL = Server.class.getClassLoader();

ClassLoader ee11CL = new URLClassLoader(serverCL, new URL[] { ... });

ContextHandler ee11Ctx = reflectionMagic(ee11CL);

server.setHandler(ee11Ctx);

server.start();
```

There is no need to create the `Environment` explicitly, and the ee11 environment will have the right ClassLoader.
Previously, it would have had the wrong ClassLoader.

gregw committed:
updates from review

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/jetty-12.1.x/12853-WebAppClassLoaderTest-ConcurrentModificationException

# Conflicts:
#	jetty-ee10/jetty-ee10-tests/jetty-ee10-test-integration/src/test/resources/RFC2616Base.xml
#	jetty-ee11/jetty-ee11-tests/jetty-ee11-test-integration/src/test/resources/RFC2616Base.xml
#	jetty-ee9/jetty-ee9-tests/jetty-ee9-test-integration/src/test/resources/RFC2616Base.xml

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/jetty-12.1.x/12853-WebAppClassLoaderTest-ConcurrentModificationException

gregw committed:
fixed XML

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/jetty-12.1.x/12853-WebAppClassLoaderTest-ConcurrentModificationException

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/jetty-12.1.x/12853-WebAppClassLoaderTest-ConcurrentModificationException

olamy commented on jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentManagerTest.java:
```diff
@@ -35,17 +39,19 @@
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
 @ExtendWith(WorkDirExtension.class)
+@Isolated
```
we might not need this anymore if the code is now thread-safe?
